keps/2941-DRA/README.md

```diff
@@ -0,0 +1,754 @@
+queuing, quota management and preemptable workloads for cluster users.
+
+### Notes/Constraints/Caveats (Optional)
+
+- The `ResourceClaims` and `ResourceClaimTemplates` APIs for DRA in k8s are immutable.
+- Device classes mapping to multiple resource names can create ambiguity during quota enforcement and requires careful
+  handling.
+- This design does not work with Kueue's Topology Aware Scheduling feature and will be addressed in future work.
+- This implementation focuses on structured parameters (beta in 1.32 and 1.33) and does not support alpha DRA features
+  like DRADeviceTaints, DRAAdminAccess, DRAPrioritizedLists, and DRAPartitionableDevices.
+
+### Risks and Mitigations
+
+With newer DRA features like DRAAdminAccess and DRAPrioritizedLists, there is a risk that
+effective tallying of resources will not be available until after allocation of these resources by kube-scheduler.
+
+In order to mitigate this risk, Kueue can take the following approach:
+1. For DRAPrioritizedLists: all the mentioned device classes in the request will be counted against the quota
+2. For DRAAdminAccess: This feature can only be enabled in admin namespace, therefore it should be skipped for being
+   counted against ClusterQuota. This is a different stance than Kubernetes ResourceQuota because kubernetes
+   ResourceQuota is namespace scoped. As a result, admin users can account for quota independent of user workloads.
+   On the contrary, with Kueue, since quota is part of ClusterQuota a cluster scoped object, admin workloads using
+   devices in admin namespace if counted against quota, will eat up quota meant for user workloads.
+3. For ResourceClaims with allocation mode `All`: worst-case scenario of the max number of devices that could be
+   allocated to a single claim will be used against quota.
+
+
+## Design Details
+
+A new feature gate DynamicResourceAllocation will be introduced, allowing users to test it in dev environments while
+making the changes dormant for production users. The following sections will explain the design in detail.
+
+### DynamicResourceAllocationConfig API
+
+```golang
+// DynamicResourceAllocationConfig is a singleton CRD that maps resource names todevice classes
+// used in ClusterQueue resource quotas. It is singleton as "default" is the only allowed named for the CRD instance in
+// Kueue namespace.
+type DynamicResourceAllocationConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of DynamicResourceAllocationConfig
+    Spec DynamicResourceAllocationConfigSpec `json:"spec"`
+}
+
+// DynamicResourceAllocationConfigSpec defines the mappings between resource names and device classes
+type DynamicResourceAllocationConfigSpec struct {
+    // Resources is a list of mappings from resource name to device classes
+    // +listType=atomic
+    // +kubebuilder:validation:MaxItems=16
+    Resources []Resources `json:"resources"`
```

Comment:
```suggestion
    Resources []DynamicResources `json:"resources"`
```
or DRAResources

Reply:
https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129

I thought @tenzen-y wanted alignment in field name and struct name.

Reply:
https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129

actually suggests DanamIcResource as the struct name, which is even better. 

Generally struct names can be more specific as they are global. Field names can be shorter and rely on the context.

Reply:
> https://github.com/kubernetes-sigs/kueue/pull/5149#discussion_r2161883129
> 
> I thought @tenzen-y wanted alignment in field name and struct name.

When I left the comment, the field name was different. So, the comment has already been outdated.
As @mimowo mentioned, we want to use the API specific name as a Go API prefix to avoid unexpected conflicts with other object.